walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  public static final Comparator<Event> SORT_BY_START = new Comparator<Event>() {
+    @Override
+    public int compare(Event a, Event b) {
+      return Long.compare(a.getWhen().start(), b.getWhen().start());
+    }
+  };
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) { 
+      Collection<TimeRange> expected = Arrays.asList();
+      return expected;
+    } else if (events.isEmpty() || request.getAttendees().isEmpty()) {
+      Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+      return expected;
+    } else {
+      // sort events by start time
+      List<Event> eventsList = new ArrayList<Event>();
+      for (Event event : events) {
+        if (request.getAttendees().containsAll(event.getAttendees())) {
+          eventsList.add(event);
+        }
+      }
+      Collections.sort(eventsList, SORT_BY_START);
+      
+      //ignore people not attending
+      if (eventsList.isEmpty()) {
+        Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+        return expected;
+      }
+
+      //merge event times
+      List<TimeRange> mergedEventsList = new ArrayList<TimeRange>();
+      if (eventsList.size() >= 2) {
+        for (int i = 0; i < eventsList.size()-1; i++) {
+          Event event1 = eventsList.get(i);
+          Event event2 = eventsList.get(i+1);
+          if (event1.getWhen().contains(event2.getWhen())) {
+            mergedEventsList.add(event1.getWhen());
+          } else if (event1.getWhen().overlaps(event2.getWhen())) {
+            mergedEventsList.add(TimeRange.fromStartEnd(event1.getWhen().start(), event2.getWhen().end(), false));
+          } else {
+            mergedEventsList.add(event1.getWhen());
+            mergedEventsList.add(event2.getWhen());
```

Comment:
Will this add duplicates of `eventList.get(i+1)` if `eventList.get(i+1)` and `eventList.get(i+2)` don't overlap?

Reply:
If it does, is there a different way to see if the event we're holding overlaps the last event we saw?

Reply:
the overlap is accounted for when we creating available times, as we will check if the duration between one event's end and the next event's start is >= requested meeting duration. If they overlap, then that difference will be <= 0, which will always be less than the requested meeting duration.